### PR TITLE
adding support for and now pinning on uphold intermediate certificate

### PR DESCRIPTION
**Pull request by husobee**
### Summary

Change to use the intermediate certificate for validation instead of the leaf certificate.

fixes: #470 

### Type of change ( select one )

- [ ] Product feature
- [x] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [x] Does your code build cleanly without any errors or warnings?
- [x] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [x] Is there a clear title that explains what the PR does?
- [x] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [x] Have you performed a self review of this PR?

### CODE REVIEW

File: utils/pindialer/dialer.go [hunk 1]
```diff
@@ -21,11 +21,13 @@ func MakeDialer(fingerprint string) Dialer {
 		}
 		connstate := c.ConnectionState()
 		for _, chain := range connstate.VerifiedChains {
-			leafCert := chain[0]
-			hash := sha256.Sum256(leafCert.RawSubjectPublicKeyInfo)
-			digest := base64.StdEncoding.EncodeToString(hash[:])
-			if digest == fingerprint {
-				return c, nil
+			// allow for intermediate certificate pinning, or leaf certificate pinning
+			for _, node := range chain {
```
Comment (corymcdonald) [inline utils/pindialer/dialer.go:25]: Since we're checking the chain, is there something we need to do to ensure we don't validate expired certs?

**Review by corymcdonald — DISMISSED** [review]

File: utils/wallet/provider/uphold/uphold.go [hunk 2]
```diff
@@ -55,8 +55,12 @@ const (
 )
 
 const (
-	sandboxFingerprint = "d961PqM/qoU9+Oc1t0v9V76A9ZlB5Dx/RFnhi4EkLIA="
-	prodFingerprint    = "d961PqM/qoU9+Oc1t0v9V76A9ZlB5Dx/RFnhi4EkLIA="
+	// The Leaf Certificates
+	//sandboxFingerprint = "d961PqM/qoU9+Oc1t0v9V76A9ZlB5Dx/RFnhi4EkLIA="
```
Comment (evq) [inline utils/wallet/provider/uphold/uphold.go:59]: We can just delete these, if we need them again we can lookup or grab from git history

**Comment by evq** [comment]
```
[user@work:~/source/bat-go]% go run bin/get-cert-fingerprint/main.go api.uphold.com:443
Dialing api.uphold.com:443
Issuer: CN=Cloudflare Inc ECC CA-3,O=Cloudflare\, Inc.,C=US Fingerprint: d961PqM/qoU9+Oc1t0v9V76A9ZlB5Dx/RFnhi4EkLIA=
Issuer: CN=Baltimore CyberTrust Root,OU=CyberTrust,O=Baltimore,C=IE Fingerprint: Y9mvm0exBk1JoQ57f9Vm28jKo5lFm/woKcVxrYxu80o=
[user@work:~/source/bat-go]% go run bin/get-cert-fingerprint/main.go api-sandbox.uphold.com:443
Dialing api-sandbox.uphold.com:443
Issuer: CN=Cloudflare Inc ECC CA-3,O=Cloudflare\, Inc.,C=US Fingerprint: d961PqM/qoU9+Oc1t0v9V76A9ZlB5Dx/RFnhi4EkLIA=
Issuer: CN=Baltimore CyberTrust Root,OU=CyberTrust,O=Baltimore,C=IE Fingerprint: Y9mvm0exBk1JoQ57f9Vm28jKo5lFm/woKcVxrYxu80o=
```

Verified new values, this PR will pin the intermediate certificate issued by CyberTrust to Cloudflare.

File: utils/pindialer/dialer.go (hunk 1 above)
Comment (husobee) [reply utils/pindialer/dialer.go:25]: The tls.Dial takes care of checking for expired certificates, and other validity checks.  We only care that the particular certificate that we know is good is in the verified chain.

**Comment by husobee** [comment]
https://github.com/brave/security/issues/223

**Review by corymcdonald — APPROVED** [review]

**Review by evq — APPROVED** [review]

**Pull request merged**